pkg/apis/storage/validation/validation.go

```diff
@@ -165,8 +166,14 @@ func validateAttacher(attacher string, fldPath *field.Path) field.ErrorList {
 // validateSource tests if the source is valid for VolumeAttachment.
 func validateVolumeAttachmentSource(source *storage.VolumeAttachmentSource, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath, ""))
+	pvNameNotExist := source.PersistentVolumeName == nil || len(*source.PersistentVolumeName) == 0
+	if source.InlineVolumeSource == nil && pvNameNotExist {
+		allErrs = append(allErrs, field.Required(fldPath, "must specify exactly one of InlineVolumeSource and PersistentVolumeName"))
+	}
+	if source.InlineVolumeSource != nil && !pvNameNotExist {
```

Comment:
nit: the double-negative is hard to follow... suggest separating the mutually-exclusive validation checks from the validation of InlineVolumeSource:

```go
switch {
case source.InlineVolumeSource == nil && source.PersistentVolumeName == nil:
  ... error about exactly one being required

case source.InlineVolumeSource != nil && source.PersistentVolumeName != nil:
  ... error about exactly one being required

case source.PersistentVolumeName != nil:
  if len(*source.PersistentVolumeName) == 0 {
    ... invalid error
  }

case source.InlineVolumeSource != nil:
  ... call storagevalidation.ValidateCSIPersistentVolumeSource
}
```

Reply:
ah. The logic was even wrong originally too, thanks for the suggestion